lib/puppet/etc.rb

```diff
@@ -0,0 +1,118 @@
+    def getgrent
+      if group_entry = Etc.getgrent
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # closes handle to /etc/group file
+    def endgrent
+      Etc.endgrent
+    end
+
+    # effectively equivalent to IO#rewind of /etc/group
+    def setgrent
+      Etc.setgrent
+    end
+
+    # Etc::getpwent returns an Etc::Passwd struct object
+    # On first call opens /etc/passwd and returns parse of first entry. Each subsquent call
+    # returns new struct for the next entry or nil if EOF. Call ::endgrent to close file.
+    def getpwent
+      if user_entry = Etc.getpwent
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    # closes handle to /etc/passwd file
+    def endpwent
+      Etc.endpwent
+    end
+
+    #effectively equivalent to IO#rewind of /etc/passwd
+    def setpwent
+      Etc.setpwent
+    end
+
+    # Etc::getpwnam searches /etc/passwd file for an entry corresponding to
+    # username.
+    # returns an Etc::Passwd struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getpwnam(username)
+      if user_entry = Etc.getpwnam(username)
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    # Etc::getgrnam searches /etc/group file for an entry corresponding to groupname.
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrnam(groupname)
+      if group_entry = Etc.getgrnam(groupname)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getgrid searches /etc/group file for an entry corresponding to id.
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrgid(id)
+      if group_entry = Etc.getgrgid(id)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getpwuid searches /etc/passwd file for an entry corresponding to id.
+    # returns an Etc::Passwd struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getpwuid(id)
+      if user_entry = Etc.getpwuid(id)
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    private
+    # utility method for converting the values of a struct returned by the Etc
+    # module to UTF-8.
+    #
+    # Warning! This is a destructive method - the struct passed is modified!
+    #
+    # Structs returned by the ruby Etc module contain members with
+    # fields of type String, Integer, or Array of Strings, so we handle these
+    # types. Otherwise ignore fields.
+    def convert_field_values_to_utf8!(struct)
+      struct.each_with_index do |value, index|
+        if value.is_a?(String)
+          struct[index] = Puppet::Util::CharacterEncoding.convert_to_utf_8(value)
```

Comment:
Thanks @joshcooper. It wasn't clear to me the implications of this, but I see that logic - it would be confusing for a user to suddenly encounter failures for properties of resources that aren't managed at all. Despite @branan's saying just that I hadn't connected the dots.

The beauty of the approach in this PR was its position at a single point of ingress for a ton of data that enters as `is` values during a transaction. Unfortunately, the actual handling and manipulation of the specific properties under management (subsets of any given `Etc` struct) is spread out across myriad providers in a web of meta-programming and obfuscation (see `nameservice` and `useradd` providers), so scope expands quite a bit if we're going to fix this there (we've been trying to do this, and it's been error-prone whack-a-mole).

For the purposes of this PR, what if instead of failing, `Puppet::Etc` is augmented with a (non-!) method that warns on the inability to convert the string, and just leaves the original string intact? In some cases, this is just going to fail upstream in the provider, but then at least a) its only failing the managed resource and b) its the same failure that was happening before. Thoughts?